File: hub/hub/schemas/questions.py
"""Question schemas."""

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field


class QuestionOption(BaseModel):
    """One offered answer.

    `label` is what comes back when it is chosen; `description` is what lets an operator pick
    without already knowing the trade-off, and is the whole reason options beat a text box.
    """

    label: str = Field(min_length=1, max_length=200)
    description: str = Field(default="", max_length=500)

    model_config = {"extra": "forbid"}


class QuestionCreate(BaseModel):
    from_agent: str = Field(max_length=64)
    question: str = Field(max_length=10000)
    blocking: bool = False
    # Offered answers. Empty means open-ended. Capped so one question cannot render an
    # unbounded wall of buttons in front of an operator deciding under a run's timeout.
    # Required, not optional. Teaching an agent to pass options is probabilistic — it forgets,
    # and the operator gets a text box for a decision that was always a choice. A schema it
    # cannot satisfy without them is not. This mirrors Claude Code's own AskUserQuestion, where
    # header/options/multiSelect are all mandatory and free text is the UI's escape, never a
    # missing-options case.
    options: List[QuestionOption] = Field(min_length=2, max_length=8)
    header: str = Field(min_length=1, max_length=64)
    multi_select: bool


class QuestionAnswer(BaseModel):
    """The operator's answer.

    `answer` is always the human-readable form. `labels` carries the structure when options were
    chosen, so a multi-select answer stays a list rather than a string someone has to re-split.
    """

    answer: str = Field(max_length=10000)
    labels: List[str] = Field(default_factory=list, max_length=8)


class QuestionResponse(BaseModel):
    id: str = Field(max_length=128)
    project_id: str = Field(max_length=128)
    from_agent: str = Field(max_length=64)
    question: str = Field(max_length=10000)
    answer: Optional[str] = Field(default=None, max_length=10000)
    answered: bool
    blocking: bool
    options: List[QuestionOption] = Field(default_factory=list)
    header: Optional[str] = None
    multi_select: bool = False

    answer_labels: List[str] = Field(default_factory=list)
    # Batch identity. A question asked on its own reports `batch_size == 1`, so a reader never has
    # to special-case the unbatched form. `batch_size` is what lets the panel say "2 of 3" while
    # holding only the unanswered rows.
    batch_id: Optional[str] = None
    batch_index: int = 0
    batch_size: int = 1
    created_at: datetime
    answered_at: Optional[datetime] = None
    # The operator closed this without answering. Distinct from `answered` on purpose: an empty
    # answer says they responded with nothing, this says they chose not to respond
    # (`2026-08-11-declining-a-question`, D1).
    declined: bool = False
    declined_at: Optional[datetime] = None
    # Whether anyone is still waiting on this question.
    #
    # Computed from the asking run rather than stored: the `runs` table already holds it, and a
    # stored copy would go stale at exactly the transition it exists to describe — the run ending.
    #
    # Defaults to `True` when the asking run is unknown, matching the answer path's presumption.
    # Guessing the other way would mark a live question inert and sort it behind dead ones, which is
    # the worse error of the two.
    asker_waiting: bool = True

    model_config = {"from_attributes": True}

File: hub/ui/src/api/questions.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { getJson, patchJson, postJson } from './client'
import { useConfigStore } from '@/store/configStore'

export interface QuestionOption {
  label: string
  /** What choosing this actually means — the reason options beat a bare text box. */
  description: string
}

export interface Question {
  id: string
  project_id: string
  from_agent: string
  question: string
  blocking: boolean
  /** Answers the agent offered. Empty means open-ended; the operator may always type instead. */
  options?: QuestionOption[]
  /** Short chip naming the decision, e.g. "Database". */
  header?: string | null
  multi_select?: boolean
  answer_labels?: string[]
  answered: boolean
  answer?: string
  /** Batch identity. One `ask_user` call can carry several questions; a question asked on its
   *  own reports `batch_size: 1`, so nothing has to special-case the unbatched form. */
  batch_id?: string | null
  batch_index?: number
  batch_size?: number
  created_at: string
  answered_at?: string
  /** The operator closed this without answering. Distinct from `answered`: an empty answer says
   *  they responded with nothing, this says they chose not to respond. */
  declined?: boolean
  declined_at?: string | null
  /** Whether anyone is still waiting on this. False once the asking run has ended — its `ask_user`
   *  call is gone, so nothing receives an answer as a result any more. Absent means "assume yes",
   *  which matches the Hub's own presumption for a question with no recorded asker. */
  asker_waiting?: boolean
}

/** How long a run waits for an answer when its agent sets no `question_timeout_seconds`.
 *
 * Mirrors `QUESTION_ANSWER_TIMEOUT` in `hub/hub/mcp_server.py` — 240s, what an ordinary MCP tool
 * call was measured tolerating. Restated here because nothing on the wire carries it: the
 * questions payload has no timeout field at all, and the roster reports `null` to mean "the
 * built-in default" without ever saying what that default is. `AgentSettingsPage` already states
 * the same 240 as its placeholder, so this is the second copy, not the first.
 */
export const DEFAULT_QUESTION_TIMEOUT_SECONDS = 240

export function useQuestions(answered?: boolean) {
  const { isConfigured, selectedProjectId: projectId } = useConfigStore()
  const params = answered !== undefined ? `?answered=${answered}` : ''
  return useQuery<Question[]>({
    queryKey: ['project', projectId, 'questions', answered],
    queryFn: () => getJson<Question[]>(`/api/v1/projects/${projectId}/questions${params}`),
    enabled: isConfigured && !!projectId,
    // An agent blocks on a question it asked, so arriving late is close to not arriving. SSE
    // already invalidates this key; the interval is the backstop for a dropped event.
    refetchInterval: 3000,
  })
}

export function useAnswerQuestion() {
  const queryClient = useQueryClient()
  const { selectedProjectId: projectId } = useConfigStore()
  return useMutation({
    mutationFn: ({ id, answer, labels }: { id: string; answer: string; labels?: string[] }) =>
      patchJson<Question>(`/api/v1/projects/${projectId}/questions/${id}`, {
        answer,
        labels: labels ?? [],
      }),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['project', projectId, 'questions'] })
      queryClient.invalidateQueries({ queryKey: ['project', projectId, 'status'] })
    },
  })
}

/**
 * Close a question without answering it.
 *
 * Also invalidates tasks: a question that had parked a task as blocked releases it, so the board is
 * stale the moment this succeeds.
 */
export function useDeclineQuestion() {
  const queryClient = useQueryClient()
  const { selectedProjectId: projectId } = useConfigStore()
  return useMutation({
    mutationFn: ({ id }: { id: string }) =>
      postJson<Question>(`/api/v1/projects/${projectId}/questions/${id}/decline`),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['project', projectId, 'questions'] })
      queryClient.invalidateQueries({ queryKey: ['project', projectId, 'status'] })
      queryClient.invalidateQueries({ queryKey: ['project', projectId, 'tasks'] })
    },
  })
}
